src/State/Scope.php

```diff
@@ -130,17 +130,39 @@ public function setExtras(array $extras): self
     /**
      * Sets the given data in the user context.
      *
-     * @param array $data The data
+     * @param array $data  The data
+     * @param bool  $merge If true, $data will be merged into user context instead of replacing it
      *
      * @return $this
      */
-    public function setUser(array $data): self
+    public function setUser(array $data, bool $merge = false): self
     {
+        if ($merge) {
+            $this->user->merge($data);
+
+            return $this;
+        }
+
+        @trigger_error('Replacing user context on setUser is deprecated since sentry-php 2.3 and will be changed to a merge on 3.0. '
+             . 'For now you can use $merge = true to merge instead of replacing it.', E_USER_DEPRECATED);
+
         $this->user->replaceData($data);
 
         return $this;
     }
 
+    /**
+     * Clears all user context data.
+     *
+     * @return $this
+     */
+    public function clearUser(): self
```

Comment:
Although having the `clear` word in the method's name is fine for me, Unified API specs talks about a `removeUser` method so we should stick with it. But if @HazAT is fine with changing such specs (I don't know if any of the other SDKs implement this method) we may leave it as is or rename it to `clearUserContext` (but this would be inconsistent with the rest of the methods, which I would have named with the `Context` suffix by the way)

Reply:
I named it that way to be consistent with `Scope::clear()` and `Scope::clearBreadcrumbs()`, but yeah, we should go with the unified spec. Renamed it to removeUser(): 304803c

(I find the behaviour of clearing the context on setUser() with an empty array to be pretty cryptic)

Reply:
I had a look at the other SDKs and I read again the documentation and I think I misunderstood what was written there. Basically, the documentation suggests as a way to remove data from a context the `remove*` method. However this method should not clear the whole context, but just remove a given item from it, so can you please update the code to selectively unset a certain key of the context?

Reply:
Ref: https://docs.sentry.io/development/sdk-dev/unified-api/#scope

> scope.set_user(user): Shallow merges user configuration (email, username, …). Removing user data is SDK-defined, either with a remove_user function or by passing nothing as data.

Yeah, it's not absolutelly clear if they meant for it to remove all context or just part of the data, but look at this other spec:
> scope.set_extra(key, value): Sets the extra key to an arbitrary value, overwriting a potential previous value. Removing a key is SDK-defined, either with a remove_extra function or by passing nothing as data.

Comparing with this spec, what I get is that on methods which set a single key of context, remove* should remove a single key (worded: "Removing **a key** is SDK-defined". The specifically mention "**a key**"). On the other hand, when having methods that set the whole context at once, without a key, the remove* method should clear the whole context (worded: "Removing **user data** is SDK-defined", as opposed to "**a key**"). 

WDYT?

Reply:
We could also open an issue on the docs asking them to clarify on that topic and go without a `removeUser` method for now. I can't actually see much use for it anyways, as the scope should be entirely cleared when changing context on the application.

Reply:
I think that you are right in saying that generally speaking the specs should be more clear on how these methods should act, because in certain parts they are and in others they are not. I suggest to go with the `removeUser` method that unsets a given key and not the whole context, as it may turn useful for example if something out of the control of the user sets some information on a parent scope and without such method you would have no way to remove it because inner scopes just inherit all data. We can add a `clearUser` context if we see that people need it

Reply:
Makes sense, there you go: 744a040

Reply:
We can revisit this after getting feedback on https://github.com/getsentry/sentry-docs/issues/1407 ✌

Reply:
I'm sorry to change one more time my mind, but since the discussion in the referenced issue may last for a long time, let's drop the `removeUser` method so that we can merge this PR as-is and if something will change in the future regarding this topic we will update the SDK accordingly

Reply:
there you go 031baaf